### ISSUE TITLE
ZDoom - Game ignores slow motion, frame advance steps too far ahead 

### ISSUE THREAD

**Issue by MoskovchenkoD** [clementgallet/libTAS#478]
**Installation**: 
1. Download ZDoom 2.8.1 from here: https://zdoom.org/downloads
2. Download any standalone gamedata, so the game could be launched. Here is Freedoom, which is an open source project: https://github.com/freedoom/freedoom/releases/download/v0.12.1/freedoom-0.12.1.zip
3. Copy "freedoom2.wad" to the source port folder, so it would look like that:
<img width="454" alt="Screenshot 2022-04-23 at 18 33 22" src="https://user-images.githubusercontent.com/12684111/164912799-46ad9426-aa1c-407f-ad8b-4035a01a6ed9.png">
4. (completely optional) Install dependencies from here, if the game won't start up: https://zdoom.org/wiki/Compile_ZDoom_on_Linux#Install_dependencies 

(

**Software launch**:
Game executable: `/home/<user>/Downloads/zdoom/zdoom`
Command-line options: `-iwad freedoom2.wad -warp 01 -nomonsters` (optional, it just launches specific game, map and removes enemies, so nothing would bother you)

Useful, but optional, source port options related to framerate:
[cl_capfps](https://zdoom.org/wiki/CVARs:Display#cl_capfps) [0..1] - locks fps at 35. 1 = true
[vid_maxfps](https://zdoom.org/wiki/CVARs:Display#vid_maxfps) [35..1000] - max framerate
To [use](https://zdoom.org/wiki/CVARs#Usage) an option, add it to the end of the command-line options, like:
`-iwad freedoom2.wad -warp 01 -nomonsters +vid_maxfps 60`

**Step to reproduce**:
After the game loads, try using Frame advance and Slow motion.

**Expected**:
- slow motion should affect game speed (not important for TAS building, but still);
- frame advance should skip 1/60 sec per a key-press.

**Actual**:
- the game always runs at 100% game speed;
- frame advance skips about 0.2 secs per a key-press. Such a big time quant is still considered as a single movie frame, which results in a big difference between in-game time and movie length.
<img width="998" alt="Screenshot 2022-04-23 at 12 44 31" src="https://user-images.githubusercontent.com/12684111/164915128-bb273e9a-3bf6-475e-b4ec-855cc9f4b433.png">

**What is tried**:
No libTAS options are required to launch the game. I tried checking all options, times, ticks, native events etc., different combinations of options. It simply doesn't solve the problem.
The source port runs at 60 fps regardless of vertical sync either inside or outside the game on Linux.

OS: Ubuntu 20.04 LTS inside VirtualBox.

**Comment by clementgallet** [comment]
See this for the culprit:

https://github.com/coelckers/gzdoom/blob/1a9bc53d84b5cceb567fd8246c44984aac88388a/src/posix/sdl/i_timer.cpp#L137-L154

It is registering a signal on SIGALARM, and the handler is calling `SDL_GetTicks()`. This messes up libTAS time handling, because it can interrupt a function that has the `native` flag (meaning it has access to native function behaviour), so that the handler has access to real time.

This way of doing engine programming feels very very wrong, and it was discontinuated because code is from 2016. Do you need to use that older version?

**Comment by MoskovchenkoD** [comment]
> It is registering a signal on SIGALARM, and the handler is calling `SDL_GetTicks()`. This messes up libTAS time handling, because it can interrupt a function that has the `native` flag (meaning it has access to native function behaviour), so that the handler has access to real time.
So, do I understand right that it requires game-specific time handling in order to fix frame advance steps?

> This way of doing engine programming feels very very wrong, and it was discontinuated because code is from 2016. Do you need to use that older version?
That's a good question. 
1. Using elder versions of ZDoom or GZDoom is necessary in order to launch and complete elder Doom mods which were made for specifically those elder source port versions, because nowadays versions have big problems with backwards compatibility.
2. The actual version of GZDoom is very likely to use the same time mechanism as in ZDoom. 
3. Moreover, its current development is focused on multi-threading. At least, I know that OpenGL and Vulkan renders (and probably OpenAL) are executed on a separate CPU core, which leads to more non-determinism that libTAS has to handle. Yes, we still have Software render which is operated on the same CPU core as the rest of the source port, but most modern mods are made with OpenGL in mind.

On the other hand, we may expect a more lenient time handling behaviour in elder versions of ZDoom (which utilise SDL 1.2), if our problem relates to SDL2.

**Comment by clementgallet** [comment]
> So, do I understand right that it requires game-specific time handling in order to fix frame advance steps?

I'm not sure what it would take to fix this, libtas never expected game code to run from signal handlers.

>     1. Using elder versions of ZDoom or GZDoom is necessary in order to launch and complete elder Doom mods which were made for specifically those elder source port versions, because nowadays versions have big problems with backwards compatibility.

Could intermediate versions of GZDoom be used? (see comment below about version change)

>     2. The actual version of GZDoom is very likely to use the same time mechanism as in ZDoom.

It was changed in 2017, see https://github.com/coelckers/gzdoom/commit/307d8931d894d8559025187ef0ae3a80f1a7dd99. First version that incorporates this is 3.2.2

>     3. Moreover, its current development is focused on multi-threading. At least, I know that OpenGL and Vulkan renders (and probably OpenAL) are executed on a separate CPU core, which leads to more non-determinism that libTAS has to handle. Yes, we still have Software render which is operated on the same CPU core as the rest of the source port, but most modern mods are made with OpenGL in mind.

Maybe there are some options or tweaks for that?

> On the other hand, we may expect a more lenient time handling behaviour in elder versions of ZDoom (which utilise SDL 1.2) available for Linux, if our problem relates to SDL2.

That's not really about SDL version.

**Comment by MoskovchenkoD** [comment]
> > ```
> > 2. The actual version of GZDoom is very likely to use the same time mechanism as in ZDoom.
> > ```
> 
> It was changed in 2017, see [coelckers/gzdoom@307d893](https://github.com/coelckers/gzdoom/commit/307d8931d894d8559025187ef0ae3a80f1a7dd99). First version that incorporates this is 3.2.2

Ooh, so maybe there's still hope! I will give it a shot later today.

> 
> > ```
> > 3. Moreover, its current development is focused on multi-threading. At least, I know that OpenGL and Vulkan renders (and probably OpenAL) are executed on a separate CPU core, which leads to more non-determinism that libTAS has to handle. Yes, we still have Software render which is operated on the same CPU core as the rest of the source port, but most modern mods are made with OpenGL in mind.
> > ```
> 
> Maybe there are some options or tweaks for that?

I don't recall if the game provides an option to do such management. Maybe Ctrl+F'ing various .ini files for related flags would give some results.
Alternatively, I think TAS conception can be promoted to GZDoom devs, so such tweaks would be released in the future.

**Comment by InfoTeddy** [comment]
Is there a reason that ZDoom can't just... not use SIGALARM? Abusing signals like that for time handling is easily the worst game programming I've seen in a while, and I've seen a *lot* of terrible and bad game programming over the years. Seriously, what were they *thinking*?!

**Comment by MoskovchenkoD** [comment]
> > 
> > It was changed in 2017, see [coelckers/gzdoom@307d893](https://github.com/coelckers/gzdoom/commit/307d8931d894d8559025187ef0ae3a80f1a7dd99). First version that incorporates this is 3.2.2
> 
> Ooh, so maybe there's still hope! I will give it a shot later today.

I got no success.
GZDoom 4.7.1: cannot bypass OpenGL 3.3 requirement on VirtualBox, even if I set Software render. For personal reasons, it's the only way I can use for now.
GZDoom 3.3.0: crashes on the startup in libTAS with the following log:
`*** Fatal Error ***
[libTAS f:0] Thread 25579 (main) ERROR: send() returns -1 with error Socket operation on non-socket
[libTAS f:0] Thread 25579 (main) ERROR: send() returns -1 with error Socket operation on non-socket
[libTAS f:0] Thread 25579 (main) ERROR: send() returns -1 with error Socket operation on non-socket
[libTAS f:0] Thread 25579 (main) ERROR: recv() returns -1 with error Socket operation on non-socket
[libTAS f:0] Thread 25579 (main) ERROR: Unknown socket message -1
Segmentation fault (core dumped)
recv() returns 0 -> socket closed
The connection to the game was closed. Exiting
`

**Comment by InfoTeddy** [comment]
If there's a segfault, then you should run a debugger. Next to the 'Start' button, there's 'Launch with GDB' (or 'Launch with LLDB'); click it to launch the debugger. When the game segfaults, type `thread apply all backtrace` (at least that's how you use it in GDB, I don't know about LLDB).

**Comment by MoskovchenkoD** [comment]
- GZDoom 3.3.0 cannot properly launch without Debug -> Native events ON. It cannot render in-game menu or the actual gameplay by any means. Cannot launch OpenAL. Here's the log (might not tell anything useful): https://pastebin.com/t4nSvM7u
The good news, it doesn't ignore slow motion and Frame advance works as intended.
- ZDoom 2.7.0, the earliest available Linux build, cannot find libfluidsynth.so.1. I spent so much time resolving it that I don't see any point digging into it.
- LZDoom 3.88 (based on GZDoom 3.3.2 and uses OpenGL 2) shows blank screen and crashes of any interaction. Here's the log (it's big!): https://pastebin.com/uA3WzTC0 

I don't think I will open any issue on any of the above source ports, but you may comment on the logs, and some new hope would show up.

To summarise, frame advance in ZDoom doesn't work because the game bypasses all software restrictions. The issue can be closed.
